### PR TITLE
Fix cases of calls to `FillPSBT` errantly returning `complete=true`

### LINKED ISSUE #30077: RPC: Internal bug in `walletprocesspsbt` when non_witness_utxo is not provided and a witness signature is invalid
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

Calling `walletprocesspsbt` on a signed PSBT that only includes a `witness_utxo` (no `non_witness_utxo`) with an invalid signature produces an internal bug (`CHECK_NONFATAL`):

```
❯ bitcoin-cli walletprocesspsbt "cHNidP8BAJoCAAAAAq3g8bgaDs84DkYQ6urdGcuevx4mTtc5ImnEuw/RKiZ2AAAAAAD9////O+qpBnMQ44R69kLMIMPrviyL4ml4zSC1iA8Ec7ux/8YBAAAAAP3///8ClnsOAAAAAAAWABQ1vvbTc5DqIMQ7qVHMKjCDEWY5m+DhDQAAAAAAFgAUWIMWf8n1WEfQh3v8yh+J7jCs/a577QIAAAEBH9C6DQAAAAAAFgAUsPZp8CKeWQmHjCFZ2elo6XlxIAEAAQEflqMOAAAAAAAWABRNIMqZK+4V6Fbf76G+CxrSoMafyQEIawJHMEQCIG7dSys9CpWsVOhrVH4aPc4s2hWr0gkYAwOZDWMczlweAiAj7av+wVmv37x4BNEWygqRxeP/GUblTMYzFePeJa5x0QEhA4Uz0JSTf/5CkbMHXfV7C5q8HjcjKGH6frUU57sDbHYDAAAA"
error code: -1
error message:
Internal bug detected: FinalizeAndExtractPSBT(psbtx, mtx)
wallet/rpc/spend.cpp:1618 (operator())
Bitcoin Core v27.99.0-24572cf76816
Please report this issue here: https://github.com/bitcoin/bitcoin/issues
```

### Expected behaviour

This scenario should be handled gracefully either with an explanatory error message or by ignoring those bad signatures (the latter happens when `non_witness_utxo` is provided). There are many reasons why a previously added signature could be invalid, e.g. if new inputs were added, outputs modified, etc.

A concrete example is a Payjoin transaction, where the receiver contributes inputs of their own after the sender has signed an initial PSBT. That's the scenario I was testing when I stumbled onto this (see discussion [here](https://github.com/payjoin/rust-payjoin/issues/51#issuecomment-2083231756)).


### Steps to reproduce

I produced a minimal repro by modifying the rpc_psbt.py functional test: https://github.com/grizznaut/bitcoin/commit/0ddee75a59d33ceed11f5f8a5e59bbc46054676b). After signing the initial PSBT, it substitutes the output address and attempts to call `walletprocesspsbt` again, but fails:

```
2024-05-10T02:42:01.177000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/Users/jaad/bin/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/Users/jaad/bin/bitcoin/test/functional/rpc_psbt.py", line 601, in run_test                                                                                                                                                               self.test_utxo_conversion()                                                                                                                                                                                                                 File "/Users/jaad/bin/bitcoin/test/functional/rpc_psbt.py", line 118, in test_utxo_conversion
    signed_psbt = offline_node.walletprocesspsbt(signed_psbt_new.to_base64())
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jaad/bin/bitcoin/test/functional/test_framework/coverage.py", line 50, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                     File "/Users/jaad/bin/bitcoin/test/functional/test_framework/authproxy.py", line 129, in __call__                                                                                                                                               raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Internal bug detected: FinalizeAndExtractPSBT(psbtx, mtx)
wallet/rpc/spend.cpp:1618 (operator())
```

### Relevant log output

_No response_

### How did you obtain Bitcoin Core

Compiled from source

### What version of Bitcoin Core are you using?

master@98dd4e7

### Operating system and version

MacOS Ventura 13.6

### Machine specifications

_No response_

### PR DESCRIPTION
**Pull request by willcl-ark**
Fixes: #30077

Fix cases of calls to `FillPSBT` returning `complete=true` when it's not
the case.

This can happen when some inputs have been signed but the transaction is
subsequently modified, e.g. in the context of PayJoins.

Also fixes a related bug where a finalized hex string is attempted to be
added during `walletprocesspsbt` but a CHECK_NONFATAL causes an abort.

### CODE REVIEW

**Comment by pinheadmz** [comment]
concept ACK

Confirmed the bug on master, fixed by the patch in this PR.

I'm not sure responding to the user-provided `"finalize"` option in this way is entirely correct though. It seems to me that the `"complete"` value that we return could be inaccurate even before #28414 because we haven't called `PSBTInputSignedAndVerified()` yet. So that makes me wonder if we should be verifying in `FillPSBT()` here, instead of just checking that *something* is in the scriptsig / witness:

https://github.com/bitcoin/bitcoin/blob/fe70be537783aeadf9e3f72cad07efd66775285b/src/wallet/wallet.cpp#L2226-L2230

...because as the issue demonstrates, the `"complete"` value could be misleading.

In other words, I think `"complete"` should be false already, before the `"hex"` was added to the rpc output.

**Comment by willcl-ark** [comment]
Rebased to fix CI

File: test/functional/rpc_psbt.py [hunk 1]
```diff
@@ -68,6 +68,28 @@ def set_test_params(self):
...
+
+        utxos = wallet.listunspent(addresses=[address])
+        psbt = wallet.createpsbt([{"txid": utxos[0]["txid"], "vout": utxos[0]["vout"]}], [{wallet.getnewaddress(): 0.9999}])
+        signed_psbt = wallet.walletprocesspsbt(psbt)["psbt"]
+
+        # Modify the raw transaction by changing the output address, so the signature is no longer valid
+        signed_psbt_obj = PSBT.from_base64(signed_psbt)
+        substitute_addr = wallet.getnewaddress(address_type="bech32m")
+        raw = wallet.createrawtransaction([{"txid": utxos[0]["txid"], "vout": utxos[0]["vout"]}], [{substitute_addr: 0.9999}])
+        signed_psbt_obj.g.map[PSBT_GLOBAL_UNSIGNED_TX] = bytes.fromhex(raw)
```
Comment (pinheadmz) [inline test/functional/rpc_psbt.py:87]: There might be a simpler way to malleate the PSBT. You could at least skip the `getnewaddress` call and just hard-code something, or change one of the output amounts. Maybe only if you retouch.

**Review by pinheadmz — APPROVED** [review]
ACK f57f68ba95fc8778fe2dc755da1e631fe60592c4

Reviewed code and confirmed test fails without patch. Approach is great, one question below about how much farther to take this approach...

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK f57f68ba95fc8778fe2dc755da1e631fe60592c4
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmaECVgACgkQ5+KYS2KJ
yTq5vw//QNiwjThHf65qtM+57G6YDPWQ3zVu2s278T0aZU5ND62Hu4A9/DHqDBSQ
36A4klRON1ddIfFX91tE3GrwT4PZ51vK3uhX0xJYkZ66xkhObVZz2lCifMs+A2ng
ldEoEBT86d17fyuQLle8/VNolqXqWx9Dw5RZY3WclnO0Hex2l9xiWepel4sFzAM8
P4YUgsw6ZaEbRUjh67tZ3kLlmz5l5yBNV5NaO7rbsPeBiIqizRbGl7CQlhlxgM4g
ZF4PGUd9rk0sSnWWMHUcVA+Uskuw64bDH0keJBLUYeKpNLfRUZ5KTRSJilurayY5
M2B7veORZ4MWjGHjycbpV+hdpjpQzX5z1cyCxGqYN4gC8P7kpFVBrB6sMUvOLoku
lGTC+SKE+sxjRmgYcE1QpPMz7uYs8KTgkTHpzlbyj61oEL1SylXDtOlc7nRN2I0A
XEnV8kK/2VKF3V45dTRPUwFm+2qX1NnCEyJ53x2srg9+U3px9UxfUMAuwTmZYyOV
E/9XCDouqE63CqCzqM683lmdL+sCsfA5IHi9sqzVRP2cY8s1NOkj7QdXQO0Z8V07
ycsGpj8nI8GAq/b647aYOU9O+Ni2PGq2wdkpP8ciD0Iafp7S9MkpCjHx9J+p4YPE
TQVElxmGYgx5zIPFrdrBehZCJGX/YDqj8Zd9Ra0QYwVzYg9aqow=
=KJef
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Comment by spacebear21** [comment]
Thanks @willcl-ark for picking this up! Confirmed this fixes the CHECK_NONFATAL abort when a witness signature is invalid.

tACK f57f68ba95fc8778fe2dc755da1e631fe60592c4

File: test/functional/rpc_psbt.py [hunk 2]
```diff
@@ -68,6 +68,28 @@ def set_test_params(self):
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
+    def test_psbt_incomplete_after_invalid_modification(self):
+        self.log.info("Check that PSBT is correctly marked as incomplete after invalid modification")
+        node = self.nodes[2]
+        wallet = node.get_wallet_rpc(self.default_wallet_name)
+        address = wallet.getnewaddress(address_type="bech32m")
```
[the remark this answers is not in the document]
Comment (willcl-ark) [reply test/functional/rpc_psbt.py]: Done in 7e36dca657c66bc70b04d5b850e5a335aecfb902

File: test/functional/rpc_psbt.py [hunk 3]
```diff
@@ -68,6 +68,28 @@ def set_test_params(self):
...
+        signed_psbt = wallet.walletprocesspsbt(psbt)["psbt"]
+
+        # Modify the raw transaction by changing the output address, so the signature is no longer valid
+        signed_psbt_obj = PSBT.from_base64(signed_psbt)
+        substitute_addr = wallet.getnewaddress()
+        raw = wallet.createrawtransaction([{"txid": utxos[0]["txid"], "vout": utxos[0]["vout"]}], [{substitute_addr: 0.9999}])
+        signed_psbt_obj.g.map[PSBT_GLOBAL_UNSIGNED_TX] = bytes.fromhex(raw)
+
+        # Check that the walletprocesspsbt call succeeds but also recognizes that the transaction is not complete
+        signed_psbt_incomplete = wallet.walletprocesspsbt(signed_psbt_obj.to_base64(), finalize=False)
```
Comment (furszy) [inline test/functional/rpc_psbt.py:90]: Seems that the PSBT 'outputs' field retains the previous output derivation path after replacement. `walletprocesspsbt` appends the new information without removing/updating the previous one. See https://github.com/furszy/bitcoin-core/commit/698e090ead3d0603146c98bd2ec06e44389f34cc.

Comment (willcl-ark) [reply test/functional/rpc_psbt.py:90]: Thanks @furszy, this indeed feels unclean to me.

I made a few changes in another test branch and I wonder if they make sense to you: https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:walletprocesspsbt-no-finalize-fixups

I took inspiration from https://github.com/furszy/bitcoin-core/commit/698e090ead3d0603146c98bd2ec06e44389f34cc (thanks) but it seems that I can probably just clear the entire output map as done in [L94](https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:walletprocesspsbt-no-finalize-fixups#diff-4b5e3b4b4ce6266324a6b660f376d57e002ef8b09d8285238753b9ce7b685beaR94) and then update the `PSBT_GLOBAL_UNSIGNED_TX` in the global map before calling `walletprocesspsbt` again.

Comment (furszy) [reply test/functional/rpc_psbt.py:90]: > I made a few changes in another test branch and I wonder if they make sense to you: https://github.com/bitcoin/bitcoin/compare/master...willcl-ark:bitcoin:walletprocesspsbt-no-finalize-fixups

The question here is whether walletprocesspsbt should update the existing PSBT 'outputs' information (similar to how it updates the transaction inputs) or should only append data to it as it is currently doing. In the current test form, the first entry of the PSBT 'outputs' field ends up with two BIP32 derivation paths that cannot coexist: they share the same master key fingerprint and are under the same derivation path: the first one is m/.../15 while the other one is m/.../16.

Clearing the map on the test side hides this topic under the carpet (which is also a possibility if you want to leave it for a follow-up too).

**Review by BrandonOdiwuor — COMMENTED** [review]
Concept ACK

**Review by ismaelsadeeq — COMMENTED** [review]
Tested ACK 7e36dca657c66bc70b04d5b850e5a335aecfb902

I've verified that this PR fixes the issue #30077,  `complete=false` is returned  instead of throwing an error.

Just a single test question and a non-blocking nit.

**Review by pinheadmz — APPROVED** [review]
re-ACK 7e36dca657c66bc70b04d5b850e5a335aecfb902

Changes since last ACK are minimal, slight simplification in tests to enable more legacy compatability

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 7e36dca657c66bc70b04d5b850e5a335aecfb902
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmaRgqoACgkQ5+KYS2KJ
yTq5NA/+NL1fOaCfl+P9yDIsXebxam/8Adf6iH5WUBt2TvUju06+305KL1dwvO6O
n0j8wmAeaJQe9tevuSTFYopxRaH+7uVQmrMDxKucURgBZ5I/VdBWk8v3+CPUsnKQ
v/hyt4RPMsL12rcti936UqKi77nfRc6E+7zsDOtRU0Rf6ME222lCaWUpqAjm4vPm
3nN18eXw6Az9rKHwG3KFzXxMy/bthlI6bsRtvbUqtFc6PkZ7xR7f79EauawCkJWS
vCr4RXVcrC+1FMB8MHMcbLiNTmNjGApg48PAYnaoJsolpGQoRaHRNwKeRo2w0Uvw
tk3uXeVNfwxRbYwzBI6pZ/JYTSa3u31qsVGqx3mVVxm8y1F/sA+fwsRB+AqToDaj
Kzf4P8FLkyYT1t83i1n5yRKawHrQTVcsMVhxsN/+4nqBbvdGba6E5SpXG3fBJ3sI
L/fJocldQmO9maY+I9CgjJqmM9GauMvhjSq+g0PnWE9pRUnt1n0bFy7DwdonzRFK
bXtFx2xcWxtrHSAvLZt6oxTXTqwfh5XiFMHwG93ULySpzOJTHB+lTED1TKf4gwhk
QcaVW4Oi0lQ8QE6rtq+o0nx0FFz+lA9RoT2MFWQU2L+HFGqokwI9PQTVWxzm9Vjd
ZWOF04PI0QvCz0DITFHV0HTX1YsyQobWWWXRxwttqy4GDUi0PPw=
=DAZD
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: test/functional/rpc_psbt.py (hunk 3 above)
Comment (achow101) [reply test/functional/rpc_psbt.py:90]: All of our PSBT updating operations should never remove any data from the PSBT. If the user decides to do something that invalidates a field in the PSBT, it's their duty to fix it themselves.

**Comment by achow101** [comment]
ACK 7e36dca657c66bc70b04d5b850e5a335aecfb902ACK 7e36dca657c66bc70b04d5b850e5a335aecfb902

**Comment by fanquake** [comment]
Backported to 27.x in #30467.

**Pull request merged**